### PR TITLE
fix4448

### LINKED ISSUE #4448: Conversion depth image into ros2 sensor_msgs::msg::Image has redundant timecost.
Image transportation latency is important, especially depth image for mapping, path planning, obstacle avoidance and other real-time vision algorithms.
In ROS2 package, time of conversion depth image into ros2 image can be optimized.
(1) https://github.com/microsoft/AirSim/blob/efa18738ac20f9b7433c489cb57dd16e8343a475/ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp#L1276 one-by-one by at(i,j) is slowest in all of cv::Mat access functions.
(2) There are two deep copies from ImageResponse to sensor_msgs::msg::Image. One is (1)'s one-by-one access in manual_decode_depth, the other is  from cv_bridge::CvImage.
For sake of real-time, I suggest following codes  instead in get_depth_img_msg_from_response function.
#
std::shared_ptr<sensor_msgs::msg::Image> AirsimROSWrapper::get_depth_img_msg_from_response(const ImageResponse& img_response, const rclcpp::Time curr_ros_time, const std::string frame_id)
{
    unused(curr_ros_time);
    // todo using img_response.image_data_float direclty as done get_img_msg_from_response() throws an error,
    //// hence the dependency on opencv and cv_bridge. however, this is an extremely fast op, so no big deal.
    //cv::Mat depth_img = manual_decode_depth(img_response);
    //std::shared_ptr<sensor_msgs::msg::Image> depth_img_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "32FC1", depth_img).toImageMsg();
    //depth_img_msg->header.stamp = airsim_timestamp_to_ros(img_response.time_stamp);
    //depth_img_msg->header.frame_id = frame_id;
    auto depth_img_msg = std::make_shared<sensor_msgs::msg::Image>();
    depth_img_msg->header.stamp = rclcpp::Time(img_response.time_stamp);
    depth_img_msg->header.frame_id = frame_id;
    depth_img_msg->width = img_response.width;
    depth_img_msg->height = img_response.height;
    depth_img_msg->data.resize(img_response.image_data_float.size() * sizeof(float));
	memcpy(depth_img_msg->data.data(), img_response.image_data_float.data(), depth_img_msg->data.size());
    depth_img_msg->encoding = "32FC1";
    depth_img_msg->step = img_response.image_data_uint8.size() / img_response.height;
    depth_img_msg->is_bigendian = 0;

    return depth_img_msg;
}
#
It has only one deep copy and  only call memcpy once to complete.

### PR DESCRIPTION
**Pull request by dzywater**
fix #4448

### CODE REVIEW

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 1]
```diff
@@ -1285,10 +1274,16 @@ std::shared_ptr<sensor_msgs::msg::Image> AirsimROSWrapper::get_depth_img_msg_fro
...
+    depth_img_msg->header.frame_id = frame_id;
+    depth_img_msg->width = img_response.width;
+    depth_img_msg->height = img_response.height;
+    depth_img_msg->data.resize(img_response.image_data_float.size() * sizeof(float));
+    memcpy(depth_img_msg->data.data(), img_response.image_data_float.data(), depth_img_msg->data.size());
+    depth_img_msg->encoding = "32FC1";
+    depth_img_msg->step = img_response.image_data_uint8.size() / img_response.height;
+    depth_img_msg->is_bigendian = 0;
     depth_img_msg->header.stamp = rclcpp::Time(img_response.time_stamp);
     depth_img_msg->header.frame_id = frame_id;
```
Comment (rajat2004) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1288]: Repeated, can be removed

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 2]
```diff
@@ -1285,10 +1274,16 @@ std::shared_ptr<sensor_msgs::msg::Image> AirsimROSWrapper::get_depth_img_msg_fro
...
-    std::shared_ptr<sensor_msgs::msg::Image> depth_img_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "32FC1", depth_img).toImageMsg();
+    auto depth_img_msg = std::make_shared<sensor_msgs::msg::Image>();
+    depth_img_msg->header.stamp = rclcpp::Time(img_response.time_stamp);
+    depth_img_msg->header.frame_id = frame_id;
+    depth_img_msg->width = img_response.width;
+    depth_img_msg->height = img_response.height;
+    depth_img_msg->data.resize(img_response.image_data_float.size() * sizeof(float));
+    memcpy(depth_img_msg->data.data(), img_response.image_data_float.data(), depth_img_msg->data.size());
+    depth_img_msg->encoding = "32FC1";
+    depth_img_msg->step = img_response.image_data_uint8.size() / img_response.height;
```
Comment (rajat2004) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1285]: Has this been verified to work correctly? When float data is passed, the `image_data_uint8` vector would be empty, and the `step` value would be incorrect

Comment (rajat2004) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1285]: So the step field is 0 then? Just wanted to confirm whether that's incorrect and should be fixed or not

Comment (dzywater) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1285]: > So the step field is 0 then? Just wanted to confirm whether that's incorrect and should be fixed or not
OK.
I know what you mean.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 1 above)
Comment (zimmy87) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1286]: resolved in latest commit

**Comment by zimmy87** [comment]
Thank you for the submission @dzywater! Tested this locally and it's working for me, so I am moving ahead with merging this.

**Pull request merged**